### ISSUE TITLE
grpcio-tools on FreeBSD

### ISSUE THREAD

**Issue by tundratim** [grpc/grpc#9721]
Trying to pip install grpcio-tools on a FreeBSD 10.3-STABLE system and seeing this error:

python/grpcio -Iinclude -I. -Ithird_party/boringssl/include -Ithird_party/zlib -I/usr/local/include/python2.7 -c src/core/lib/tsi/ssl_tr
p.freebsd-10.3-STABLE-amd64-2.7/src/core/lib/tsi/ssl_transport_security.o -pthread -std=gnu99
    src/core/lib/tsi/ssl_transport_security.c:350:14: error: use of undeclared identifier 'AF_INET'
            af = AF_INET;
                 ^
    src/core/lib/tsi/ssl_transport_security.c:352:14: error: use of undeclared identifier 'AF_INET6'
            af = AF_INET6;
                 ^
    2 errors generated.

Any thoughts on a resolution or workaround?

**Comment by nathanielmanistaatgoogle** [comment]
With what version of `pip` are you working and what version of `grpcio-tools` are you trying to install? Are you able to reproduce the obstacle in a [virtual environment](https://virtualenv.pypa.io/en/stable/) or only with your system `pip`?

**Comment by tundratim** [comment]
FreeBSD 10.3-STABLE #22 r313193

python 2.7.13

pip 9.0.1

grpc 0.3.post19  (installed ok)

grpcio-tools 1.1.0  (failing)

All done within a 'pew' virtual environment though I can make it fail the same way on the host OS.

**Comment by nathanielmanistaatgoogle** [comment]
The current state of gRPC-on-BSD is "well we know of some folks who tweaked a few things and told us it was working about twelve months ago", so we suspect that it's highly likely that gRPC doesn't build on BSD.

Are you able to work with gRPC from source? If so, do you see the same problem when you build and run the test suite?

**Comment by tundratim** [comment]
It would seem that a build from source also blows up - gmake + clang:

[C]       Compiling src/core/lib/iomgr/tcp_posix.c
src/core/lib/iomgr/tcp_posix.c:219:25: error: implicit conversion loses integer precision: 'msg_iovlen_type' (aka 'unsigned long') to
      'int' [-Werror,-Wshorten-64-to-32]
  msg.msg_iovlen = tcp->iov_size;
                 ~ ~~~~~^~~~~~~~
src/core/lib/iomgr/tcp_posix.c:364:22: error: implicit conversion loses integer precision: 'msg_iovlen_type' (aka 'unsigned long') to
      'int' [-Werror,-Wshorten-64-to-32]
    msg.msg_iovlen = iov_size;
                   ~ ^~~~~~~~
2 errors generated.
gmake: *** [Makefile:2307: /usr1/shared/grpc/objs/opt/src/core/lib/iomgr/tcp_posix.o] Error 1

**Comment by xorander00** [comment]
I commented on a thread in regards to this over at GitLab:
https://gitlab.com/gitlab-org/gitaly/issues/154

Figured I'd link it here too, as I'd like to get this fixed and working on FreeBSD.  Any suggestions would be greatly appreciated, thanks! :)

**Comment by tundratim** [comment]
@xorander00 

Yes, it would be good to figure this out.  I've poked a bit at the source, but not enough to get anywhere useful.

**Comment by xorander00** [comment]
@tundratim 
I did get it building with CMake after making some adjustments.  I remember finding some changes on github that allowed CMake to find the FreeBSD package for protobufs.  I deleted those changes though and need to find them again.  When I do, I'll keep better track of it and put together a patch.

My issue was after building the libraries, I didn't know how to get the Ruby gem to be built against them.

**Comment by xorander00** [comment]
Actually, I did save the build commands I used, just not the changes:

cmake .. -GNinja -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=$HOME/tmp/native -DgRPC_ZLIB_PROVIDER=package -DgRPC_SSL_PROVIDER=package -DgRPC_PROTOBUF_PROVIDER=package

$HOME/tmp/native because I wanted it to install into a clean directory so that I could examine the output structure/files.

**Comment by xorander00** [comment]
Just pinging/bumping this with a little bit of info and request for any further feedback/suggestions :)

I last left off at tackling the two places in tcp_posix.c where there's an assignment that has an implicit cast from unsigned long -> int.  I did manage to make it compile before by making it an explicit cast, but I'd never actually use that since it seems dangerous.  The compiler options -Werror and -Wshorten-64-to-32 are there for a reason I'm sure, so it's probably not safe.

Other than that, wrapping <sys/socket.h> in #ifdef GRPC_FREEBSD and patching up CMakeLists.txt to properly include any missing includes/libs for protobufs seems to allow it to compile.  It's that cast that I'm uncertain about.

**Comment by mehrdada** [comment]
I have managed to build the current grpc on master on FreeBSD 10.3 and the python package using the following script (definitely unsupported):

```bash
#!/usr/local/bin/bash

mkdir grpc-freebsd
cd grpc-freebsd
git clone https://github.com/grpc/grpc
cd grpc
git submodule update --init
sed -i \  's/set(_gRPC_ALLTARGETS_LIBRARIES dl rt m pthread)/set(_gRPC_ALLTARGETS_LIBRARIES rt m pthread)/' CMakeLists.txt
sed -i \  $'s/#include <arpa\/inet.h>/#include <arpa\/inet.h>\\\n#include <sys\/socket.h>/' src/core/tsi/ssl_transport_security.c
pushd third_party/cares
ln -s config_linux/ares_config.h
popd
cmake .
gmake
#python
virtualenv ../venv
../venv/bin/python -m pip install --upgrade pip setuptools cython
../venv/bin/python -m pip install -r requirements.txt
GRPC_PYTHON_BUILD_WITH_CYTHON=1 ../venv/bin/python -m pip install . 
```

Specifically, there were three issues to address:

1. The `src/core/tsi/ssl_transport_security.c` file does not include the appropriate headers on FreeBSD that define `AF_INET`. This was fixed by including `<sys/socket.h>`.
2. The `cares` dependency library does not have a `config_freebsd` directory so the `ares_config.h` include is not found. I could circumvent the error by symlinking `ares_config.h` to `config_linux/ares_config.h`. Makefiles need to be adjusted to include an appropriate file for FreeBSD.
3. The `CMakeLists.txt` file in root assumes every Unix is GNU/Linuxy and wants to link `dl` which fails on FreeBSD. It needs to get smarter.

**Comment by xorander00** [comment]
Yeah, that's pretty much what I did as well.  Although grpc defines GRPC_FREEBSD, so I added #ifdef & #include <sys/socket.h> into port_platform.h, since that seems to be where stuff like that should go.


I don't remember having any issues with cares, but I might have disabled it?  I don't remember off the top of my head.


Yes, the build scripts (CMake mainly) needs to be smarter in terms of that.  I saw quite a bit of stuff that made it Linux dependent.  I added a define for FreeBSD and then updated it further down to exclude dl if it detected FreeBSD.


________________________________

**Comment by mehrdada** [comment]
With the merged PRs (#10705 and $10724) and latest protobuf package, it is possible to make gRPC work on FreeBSD from source and generate code from proto files. Since it is not an officially supported platform and we don't have CI infrastructure for it, we do not have additional plans to further work in an organized fashion on this issue, so I am going to close this issue at the moment. Of course PRs to help improve the support status are always appreciated.
